### PR TITLE
Disallow legacy UPDATE/DELETE on Hive ACID tables

### PR DESCRIPTION
**Pull request by hashhar**
## Description

UDPATE/DELETE on Hive ACID transactional tables when using the non-MERGE
implementation can lead to correctness issues as shown by the tests
added to TestHiveTransactionalTable.

To prevent these issues we now disallow UPDATE/DELETE on Hive ACID
tables when using the legacy implementation (by setting
`deprecated.legacy-update-delete-implementation` to true).

## Additional context and related issues

This adds test coverage and fix for https://github.com/trinodb/trino/issues/12731

## Release notes

```markdown
# Hive

* Disallow UPDATE/DELETE on Hive ACID transactional tables to prevent correctness
  issues when `deprecated.legacy-update-delete-implementation` is set to true.
```

### CODE REVIEW

**Comment by hashhar** [comment]
The first commit was an attempt to add a toggle to disallow UPDATE/DELETE on ACID tables when using legacy implementation but it has the issue that `HiveMetadata#applyDelete` gets called in both legacy and new MERGE-based implementations. That can be seen by running `./testing/bin/ptl test run --environment singlenode-hdp3 --config config-default -- -t 'TestHiveTransactionalTable.testLargePartitioned'` and noticing that while `testLargePartitionedUpdate` passes `testLargePartitionedDelete` fails complaining it's not supported.


Since the legacy vs MERGE choice is a system session property we cannot inspect that in HiveMetadata and so I no way to selectively fail `HiveMetadata#applyDelete` only for legacy mode.

So TL;DR, ignore the first commit and instead review the entire PR as a whole which will be squashed before merge (if we merge this).

**Review by djsstarburst — COMMENTED** [review]
My understanding is that [the Trino PR that removes the legacy DELETE and UPDATE support](https://github.com/trinodb/trino/pull/15161) will be merged very soon, so I'm not sure this PR is necessary.

CC: @electrum @findepi

**Comment by hashhar** [comment]
> My understanding is that [the Trino PR that removes the legacy DELETE and UPDATE support](https://github.com/trinodb/trino/pull/15161) will be merged very soon, so I'm not sure this PR is necessary.
> 
> CC: @electrum @findepi

Agreed. However maybe we should just have the tests to guard from a future regression?

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveTransactionalTable.java [hunk 1]
```diff
@@ -2019,6 +2020,113 @@ private void unbucketedTransactionalTableWithTaskWriterCountGreaterThanOne(boole
         });
     }
 
+    @Test(groups = HIVE_TRANSACTIONAL)
+    public void testLargePartitionedDelete()
+    {
+        // Test for https://github.com/trinodb/trino/issues/12731
```
Comment (findepi) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveTransactionalTable.java:2026]: Can make it a method javadoc

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveTransactionalTable.java [hunk 2]
```diff
@@ -2019,6 +2020,113 @@ private void unbucketedTransactionalTableWithTaskWriterCountGreaterThanOne(boole
...
+        if (getHiveVersionMajor() < 3) {
+            throw new SkipException("Hive transactional tables are supported with Hive version 3 or above");
+        }
+        largePartitionedDelete("large_delete_");
+    }
+
+    @Test(groups = HIVE_TRANSACTIONAL)
+    public void testLargePartitionedLegacyDelete()
+    {
+        // Test for https://github.com/trinodb/trino/issues/12731
```
Comment (findepi) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveTransactionalTable.java:2036]: Can make it a method javadoc

**Comment by findepi** [comment]
> Agreed. However maybe we should just have the tests to guard from a future regression?

Sounds good to me.
How long do these new tests take on the CI?

**Comment by hashhar** [comment]
> > Agreed. However maybe we should just have the tests to guard from a future regression?
> 
> Sounds good to me. How long do these new tests take on the CI?

```
INFO: Test io.trino.tests.product.hive.TestHiveTransactionalTable.testLargePartitionedDelete took 41.67s
INFO: Test io.trino.tests.product.hive.TestHiveTransactionalTable.testLargePartitionedLegacyDelete took 24.7s
INFO: Test io.trino.tests.product.hive.TestHiveTransactionalTable.testLargePartitionedLegacyUpdate took 35.72s
INFO: Test io.trino.tests.product.hive.TestHiveTransactionalTable.testLargePartitionedUpdate took 1.25m
```

So around 3m. Since we plan to remove the legacy impl this would go down to ~2m.

**Review by findepi — APPROVED** [review]
Tests LGTM

"Disallow legacy DELETE/UPDATE on Hive ACID tables by default" is also a good idea, if we can make it work.

**Comment by hashhar** [comment]
> "Disallow legacy DELETE/UPDATE on Hive ACID tables by default" is also a good idea, if we can make it work.

Got it working @findepi.

**Comment by hashhar** [comment]
@findepi PTAL again.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveConfig.java [hunk 3]
```diff
@@ -1291,4 +1293,17 @@ public HiveConfig setPartitionProjectionEnabled(boolean enabledAthenaPartitionPr
...
         return this;
     }
+
+    public boolean isLegacyAcidModificationEnabled()
+    {
+        return isLegacyAcidModificationEnabled;
+    }
+
+    @Config("hive.legacy-acid-modification-enabled")
+    @ConfigDescription("Enable performing legacy DELETE/UPDATE on Hive ACID transactional tables")
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveConfig.java:1303]: We are disabling it because it has correctness issues.
Why would we have a toggle for enabling it at all?

if we keep it, the description should scream loud that you should never use it.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveSessionProperties.java [hunk 4]
```diff
@@ -550,7 +551,12 @@ public HiveSessionProperties(
                         hiveConfig.getHudiCatalogName().orElse(null),
                         // Session-level redirections configuration does not work well with views, as view body is analyzed in context
                         // of a session with properties stripped off. Thus, this property is more of a test-only, or at most POC usefulness.
-                        true));
+                        true),
+                booleanProperty(
+                        LEGACY_ACID_MODIFICATION_ENABLED,
+                        "Enable performing legacy DELETE/UPDATE on Hive ACID transactional tables",
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveSessionProperties.java:557]: As in config, remove or maye it scream out loud.

**Review by findepi — APPROVED** [review]

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveSessionProperties.java (hunk 4 above)
Comment (hashhar) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveSessionProperties.java:557]: Now throws

```
Query 20230118_094118_00005_fnwkj failed: DELETE on ACID transactional tables is not supported
when using the legacy UPDATE/DELETE implementation.Switch to the new implementation by
removing `deprecated.legacy-update-delete-implementation` system config propertyor setting
`legacy_update_delete_implementation` catalog session property to false.
```

Comment (findepi) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveSessionProperties.java]: > As in config ...

BTW https://github.com/trinodb/trino/pull/15755 would make it simpler to define (config property, session property) pairs and keep them in sync with regards to description.

Comment (hashhar) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveSessionProperties.java]: Thank you for that PR.
